### PR TITLE
[Messenger] Remove base64_encode & use addslashes

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | already covered by existing issue

In #30814, we base64_encoded messages because some transports (specifically DoctrineTransport + Postgresql & SQS) do not allow binary data.

The downside is that the messages become unreadable, which makes it much less convenient to debug your messages with 3rd party monitoring tools, for example.

This PR replaces base64_encode with addslashes. Another alternative (that I first tried in this PR) was to use a blob type, which Drupal does in its code (https://www.drupal.org/project/drupal/issues/690746). But, it still meant that binary data could cause problems with other transports, like SQS.

I also put all the serializer config under a nice, neat `serializer` key under messenger.

Best seen with `?w=1`.

Cheers!

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 1]
```diff
@@ -406,17 +406,26 @@
 
     <xsd:complexType name="messenger">
         <xsd:sequence>
-            <xsd:element name="default-serializer" type="xsd:string" minOccurs="0" />
-            <xsd:element name="symfony-serializer" type="messenger_symfony_serializer" minOccurs="0" />
-            <xsd:element name="encoder" type="xsd:string" minOccurs="0" />
-            <xsd:element name="decoder" type="xsd:string" minOccurs="0" />
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: why are the encoder and decoder gone ?

File: src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php [hunk 2]
```diff
@@ -30,10 +37,14 @@ public function decode(array $encodedEnvelope): Envelope
             throw new MessageDecodingFailedException('Encoded envelope should have at least a "body".');
         }
 
-        $serializeEnvelope = base64_decode($encodedEnvelope['body']);
+        $serializeEnvelope = $encodedEnvelope['body'];
+
+        if (true === $this->base64Encode) {
```
Comment (stof) [inline src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php]: I suggest using `if ($this->base64Encode)` as it is already a boolean anyway

**Review by nicolas-grekas — APPROVED** [review]
once stof's comments are addressed

**Comment by nicolas-grekas** [comment]
Another proposal would be to always use addslashes/stripslashes on the serialized payload.
That would allow removing the option - ie less complexity. WDYT?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: I can't find them referenced anywhere - there is no `encoder` or `decoder` in the Configuration class. I think these were incorrectly *not* removed from some earlier change to messenger.

**Comment by weaverryan** [comment]
> Another proposal would be to always use addslashes/stripslashes on the serialized payload.
That would allow removing the option - ie less complexity. WDYT?

I think this *does* fix the problem. It would not allow message objects themselves to contain binary data, but probably that's not supported? Here is what SQS says about character sets:

> A message can include only XML, JSON, and unformatted text. The following Unicode characters are allowed:

> #x9 | #xA | #xD | #x20 to #xD7FF | #xE000 to #xFFFD | #x10000 to #x10FFFF

Is addslashes sufficient?

**Comment by lyrixx** [comment]
> That would allow removing the option - ie less complexity. WDYT?

I don't like this option. I often copy/paste message from/to AMQP. More over I often have different language between my producer and my consumer. (But I'm not sure the messenger component is designed for such use cases).

---

Instead of an option, why don't you add  `Transport::needEscaping()` ?

**Comment by weaverryan** [comment]
> I often copy/paste message from/to AMQP.

What's the format you usually use? serialized PHP or JSON?

> More over I often have different language between my producer and my consumer. (But I'm not sure the messenger component is designed for such use cases).

It absolutely is. But in that case, you'd be using the Symfony serializer (JSON) or something custom. Indeed the PhpSerializer is (mostly) only for the situation where you plan to send and consume from your Symfony app.

> Instead of an option, why don't you add Transport::needEscaping() ?

That's an interesting idea. In fact, the transport is responsible for calling the serializer... and so it could even pass this as an option to `encode()` and `decode()`

**Comment by weaverryan** [comment]
> we could also and/or alternatively define a new messenger.transport.base64_encoded_php_serializer service instead of adding a new option

I like this way.

> Instead of an option, why don't you add Transport::needEscaping() ?

On 2nd thought, this might be hard to implement, as this whole escaping thing only applies if you're using the PhpSerializer... and the user may or may not be using it. So, it would be a weird transport option that only applies sometimes.

**Comment by lyrixx** [comment]
> What's the format you usually use? serialized PHP or JSON?

JSON of course :)

**Comment by weaverryan** [comment]
Ready for review again - I've updated the description, etc. This changes to use `addslashes` instead of `base64_encode()`. This only applies to the PhpSerializer.

File: src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php [hunk 3]
```diff
@@ -30,11 +30,7 @@ public function decode(array $encodedEnvelope): Envelope
             throw new MessageDecodingFailedException('Encoded envelope should have at least a "body".');
         }
 
-        $serializeEnvelope = base64_decode($encodedEnvelope['body']);
-
-        if (false === $serializeEnvelope) {
-            throw new MessageDecodingFailedException('The "body" key could not be base64 decoded.');
-        }
+        $serializeEnvelope = \stripslashes($encodedEnvelope['body']);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php]: `\` should be removed

File: src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php [hunk 4]
```diff
@@ -44,8 +40,10 @@ public function decode(array $encodedEnvelope): Envelope
      */
     public function encode(Envelope $envelope): array
     {
+        $body = \addslashes(serialize($envelope));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php]: `\` should be removed

**Review by nicolas-grekas — APPROVED** [review]
> messenger.transport.base64_encoded_php_serializer

for another PR when needed?

**Comment by weaverryan** [comment]
> messenger.transport.base64_encoded_php_serializer
> for another PR when needed?

Yep, exactly. It's trivial anyways - you could even decorate the `PhpSerializer`

Ready to go!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Pull request merged**